sequencefile/reader.go

```diff
@@ -0,0 +1,177 @@
+package sequencefile
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+)
+
+// A Reader reads key/value pairs from an input stream.
+type Reader struct {
+	Header          Header
+	syncMarkerBytes []byte
+
+	reader io.Reader
+	closed bool
+	err    error
+
+	buf   bytes.Buffer
+	key   []byte
+	value []byte
+}
+
+// New returns a new Reader for sequencefiles, reading input from r. If the
+// io.Reader is positioned at the start of a file, you should immediately call
+// ReadHeader to read through the header.
+func New(r io.Reader) *Reader {
+	return &Reader{reader: r}
+}
+
+// Scan advances the reader to the start of the next record, reading the key
+// and value into memory. These can then be obtained by calling Key and Value.
+// If the end of the file is reached, or there is an error, Scan will return
+// false.
+func (r *Reader) Scan() bool {
+	return r.scan(true)
+}
+
+// ScanKey works exactly like Scan, but only reads the key of the record into
+// memory, not the value. Subsequent calls to Value will return nil.
+func (r *Reader) ScanKey() bool {
+	return r.scan(false)
+}
+
+// Err returns the first non-EOF error reached while scanning.
+func (r *Reader) Err() error {
+	return r.err
+}
+
+// Key returns the key for the current record. The byte slice will be reused
+// after the next call to Scan.
+func (r *Reader) Key() []byte {
+	return r.key
+}
+
+// Value returns the value for the current record. The byte slice will be
+// reused after the next call to Scan.
+func (r *Reader) Value() []byte {
+	return r.value
+}
+
+func (r *Reader) scan(readValues bool) bool {
+	if r.closed {
+		return false
+	}
+
+	var b []byte
+
+	b, err := r.consume(4)
+	if err == io.EOF {
+		return false
+	} else if err != nil {
+		r.close(err)
+		return false
+	}
+
+	// Length -1 means a sync marker (obnoxiously encoded as a cast uint32).
+	totalLength := int(int32(binary.BigEndian.Uint32(b)))
```

Comment:
I think I need to to preserve the fact that hadoop writes `-1` as `int32.MAX_VALUE + 1` (but in an unsigned int). If the go runtime decided ints were 64 bits, then casting directly to `int` would lose that fact.